rasa/nlu/data_augmentation/augmenter.py

```diff
@@ -0,0 +1,736 @@
+    intents_to_augment: Set[Text],
+    augmentation_factor: Dict[Text, int],
+    random_seed: int,
+) -> TrainingData:
+    """Create randomly sampled augmentation set.
+
+       Randomly selects paraphrases for data augmentation from the generated
+       paraphrase pool.
+
+    Args:
+        nlu_training_data: NLU training data (without augmentation).
+        paraphrase_pool: Paraphrases for intents that should be augmented.
+        intents_to_augment: Intents that should be augmented.
+        augmentation_factor: Amount of data augmentation per intent.
+        random_seed: Random seed for sampling paraphrases.
+
+    Returns:
+        Paraphrases for data augmentation.
+    """
+    random.seed(random_seed)
+    new_training_data = []
+    for intent in intents_to_augment:
+        random.shuffle(paraphrase_pool[intent])
+
+        # Creates `Message` objects from the randomly shuffled list of paraphrases
+        new_training_data.extend(
+            [
+                Message(data={TEXT: item.get(TEXT), INTENT: item.get(INTENT)})
+                for item in paraphrase_pool[intent][: augmentation_factor[intent]]
+            ]
+        )
+
+    augmented_training_data = TrainingData(new_training_data).merge(nlu_training_data)
+
+    return augmented_training_data
+
+
+def _train_test_nlu_model(
+    output_directory: Text,
+    nlu_training_file: Text,
+    config: Text,
+    nlu_evaluation_data: TrainingData,
+) -> Dict[Text, Any]:
+    """Runs the NLU train/test loop using the given augmented training data.
+
+    Performs training a new NLU model with the augmented training set and subsequently
+    evaluates the model on the test data. The trained model will be stored to the given
+    output directory.
+
+    Args:
+         output_directory: Directory to store the evaluation reports and augmented
+             training data in.
+         nlu_training_file: Augmented NLU training data file.
+         config: NLU model config.
+         nlu_evaluation_data: NLU evaluation data.
+
+    Returns:
+        Classification report of the NLU model trained on the augmented training data.
+    """
+    # Train NLU model
+    model_path = train_nlu(
+        config=config, nlu_data=nlu_training_file, output=output_directory
+    )
+
+    # Load and Evaluate NLU Model
+    unpacked_model_path = get_model(model_path)
+    nlu_model_path = os.path.join(unpacked_model_path, "nlu")
+
+    interpreter = Interpreter.load(nlu_model_path)
+    interpreter.pipeline = remove_pretrained_extractors(interpreter.pipeline)
+
+    (intent_results, *_) = get_eval_data(interpreter, nlu_evaluation_data)
+    intent_report = create_intent_report(
+        intent_results=intent_results, output_directory=None, report_as_dict=True,
+    )
+    intent_errors = get_intent_errors(intent_results=intent_results)
+    rasa.shared.utils.io.dump_obj_as_json_to_file(
+        os.path.join(output_directory, "intent_errors.json"), intent_errors,
```

Comment:
how about introducing constants for `intent_errors.json` as this is used in multiple files now?

Reply:
Generally yes, and I believe I've already done this for `intent_report.json`, though I'll only replace this for occurrences within the data augmentation code. There are other occurrences (e.g. in `nlu/test.py`) that I won't replace within the scope of this PR.